src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,777 @@
+                    WriteHeaders(ref writer, m);
+                    WriteInvocationMessage(m, ref writer);
+                    break;
+                case StreamInvocationMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.StreamInvocationMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteStreamInvocationMessage(m, ref writer);
+                    break;
+                case StreamItemMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.StreamItemMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteStreamItemMessage(m, ref writer);
+                    break;
+                case CompletionMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.CompletionMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteCompletionMessage(m, ref writer);
+                    break;
+                case CancelInvocationMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.CancelInvocationMessageType);
+                    WriteHeaders(ref writer, m);
+                    WriteCancelInvocationMessage(m, ref writer);
+                    break;
+                case PingMessage _:
+                    WriteMessageType(ref writer, HubProtocolConstants.PingMessageType);
+                    break;
+                case CloseMessage m:
+                    WriteMessageType(ref writer, HubProtocolConstants.CloseMessageType);
+                    WriteCloseMessage(m, ref writer);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Unsupported message type: {message.GetType().FullName}");
+            }
+            writer.WriteEndObject();
+            writer.Flush();
+        }
+
+        private void WriteHeaders(ref Utf8JsonWriter writer, HubInvocationMessage message)
+        {
+            if (message.Headers != null && message.Headers.Count > 0)
+            {
+                writer.WriteStartObject(HeadersPropertyNameBytes);
+                foreach (var value in message.Headers)
+                {
+                    writer.WriteString(value.Key, value.Value);
+                }
+                writer.WriteEndObject();
+            }
+        }
+
+        private void WriteCompletionMessage(CompletionMessage message, ref Utf8JsonWriter writer)
+        {
+            WriteInvocationId(message, ref writer);
+            if (!string.IsNullOrEmpty(message.Error))
+            {
+                writer.WriteString(ErrorPropertyNameBytes, message.Error);
+            }
+            else if (message.HasResult)
+            {
+                var bytes = JsonSerializer.ToBytes(message.Result, message.Result?.GetType());
+                using var token = JsonDocument.Parse(bytes);
+                token.RootElement.WriteAsProperty(ResultPropertyNameBytes, ref writer);
```

Comment:
I'd also create a helper method that's called every time we write an arbitrary type, so when we fix it we don't accidentally leave the inefficient logic laying around somewhere.